### PR TITLE
separate RESTStorage by API group

### PR DESCRIPTION
**Pull request by deads2k**
This demonstrates how we could start separating the description of `RESTStorage` from `pkg/master`.  Each API group owns constructing its own RESTStorage.  Since `pkg/registry` is where we place all packages that create RESTStorage, it seems reason to split the package by `pkg/registry/<api group>/storage` to create RESTStorage and `pkg/registry/<api group>/<resource>` to support each individual resource.

The interface to construct this RESTStorage is dependent upon some `genericapiserver` types now, but we could (should?) move those items to a different location.

@kubernetes/sig-api-machinery 
@lavalamp can you weigh in on how I've split these packages?  I think we agree that we need better segregation, but we haven't spoken about how to spell it.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32802)

### CODE REVIEW

**Comment by lavalamp** [comment]
@deads2k I'm so swamped right now, is there any way you can avoid doing massive refactors for a little bit? :) :)

**Comment by deads2k** [comment]
> @deads2k I'm so swamped right now, is there any way you can avoid doing massive refactors for a little bit? :) :)

I think if you agree that `pkg/registry` is a package for building RESTStorage and each API group should have a package that knows how to build RESTStorage for its own resources, some else can take the review.

How about that and I promise that world will be closer to the way you want it when you come back to code?  Might not be exactly how you'd like it, but it will be a lot closer.

**Comment by deads2k** [comment]
> How about that and I promise that world will be closer to the way you want it when you come back to code? Might not be exactly how you'd like it, but it will be a lot closer.

@lavalamp think of it like lawn service! :)

**Comment by smarterclayton** [comment]
David's off add^H^H^Hremoving barbed wire and you try to slow him down?!

**Comment by lavalamp** [comment]
Haha. OK, I agree with the stated purpose and I think it's fine if someone else does the review.

(mostly I just selfishly want to get my api group changes in before they get harder to rebase, but I'm stuck on other things for a bit longer)

**Comment by smarterclayton** [comment]
Can someone help get the remainder of your api group changes done by
adopting it?  Someone like @sttts or @deads2k?

**Comment by lavalamp** [comment]
hm, actually it looks like #30565 might just need a retest. I have a bit
more to do after that and then help converting every group over would be
nice.

Also if someone has time to take over #31731 that would be a huge help.

**Comment by deads2k** [comment]
@derekwaynecarr my whole team is out today.  You got someone who can take a look?

File: pkg/master/master.go [hunk 1]
```diff
@@ -206,7 +201,7 @@ func New(c *Config) (*Master, error) {
 		DisableThirdPartyControllerForTesting: m.disableThirdPartyControllerForTesting,
 	}
 	c.RESTStorageProviders[policy.GroupName] = PolicyRESTStorageProvider{}
-	c.RESTStorageProviders[rbac.GroupName] = RBACRESTStorageProvider{AuthorizerRBACSuperUser: c.AuthorizerRBACSuperUser}
+	c.RESTStorageProviders[rbac.GroupName] = rbacstorage.RESTStorageProvider{AuthorizerRBACSuperUser: c.AuthorizerRBACSuperUser}
```
Comment (sttts) [inline pkg/master/master.go]: ...storage and RESTStorage is kind of redundant.

**Comment by sttts** [comment]
Having the interface in genericapiserver is ok for now. Alternative: `genericapiserver/storage` or `.../reststorage` and then `RESTStorageProvider -> Provider`.

Overall the move of the storage providers into the registry sub-packages lgtm.

**Comment by deads2k** [comment]
@k8s-bot test this: issue #32770

**Comment by deads2k** [comment]
@k8s-bot unit test this: issue #32770

**Comment by deads2k** [comment]
@k8s-bot unit test this: issue #32770

**Pull request merged**